### PR TITLE
add StreamMux and StreamDemux usage

### PR DESCRIPTION
**Pull request by Readon**
documentation for StreamMux and StreamDemux, for SpinalHDL/SpinalHDL#901 .

### CODE REVIEW

File: source/SpinalHDL/Libraries/stream.rst [hunk 1]
```diff
@@ -489,6 +489,29 @@ or
    val inputStream = Stream(Bits(8 bits))
    val outputStreams = StreamFork(inputStream,portCount=2, synchronous=true)
 
+StreamMux
+^^^^^^^^^^
+A mux implementation for stream, which connect output to input specified by `io.select` signal.
+
+.. code-block:: scala
+   val inputStreams = Vec(Stream(Bits(8 bits)), portCount)
```
Comment (numero-744) [inline source/SpinalHDL/Libraries/stream.rst:497]: Add an empty line to build a code block

```suggestion

   val inputStreams = Vec(Stream(Bits(8 bits)), portCount)
```

File: source/SpinalHDL/Libraries/stream.rst [hunk 2]
```diff
@@ -489,6 +489,29 @@ or
...
+   The `UInt` type of `select` signal could not be changed while output stream is stalled, or it might break the transaction on the fly.
+   Use `Stream` typed `select` can generate a stream interface which only fire and change the routing when it is safe.
+
+
+StreamDemux
+^^^^^^^^^^
+A demux for stream which route input to output stream specified by `select`. Refer the notes above.
+
+.. code-block:: scala
+   val inputStream = Stream(Bits(8 bits))
```
Comment (numero-744) [inline source/SpinalHDL/Libraries/stream.rst:511]:
```suggestion

   val inputStream = Stream(Bits(8 bits))
```

File: source/SpinalHDL/Libraries/stream.rst [hunk 3]
```diff
@@ -489,6 +489,29 @@ or
...
+   val select = UInt(log2Up(inputStreams.length) bits)
+   val outputStream = StreamMux(select, inputStreams)
+
+.. note::
+   The `UInt` type of `select` signal could not be changed while output stream is stalled, or it might break the transaction on the fly.
+   Use `Stream` typed `select` can generate a stream interface which only fire and change the routing when it is safe.
+
+
+StreamDemux
+^^^^^^^^^^
```
Comment (numero-744) [inline source/SpinalHDL/Libraries/stream.rst:507]: To fix a warning

```suggestion
^^^^^^^^^^^

```

File: source/SpinalHDL/Libraries/stream.rst [hunk 4]
```diff
@@ -489,6 +489,29 @@ or
    val inputStream = Stream(Bits(8 bits))
    val outputStreams = StreamFork(inputStream,portCount=2, synchronous=true)
 
+StreamMux
+^^^^^^^^^^
+A mux implementation for stream, which connect output to input specified by `io.select` signal.
```
Comment (numero-744) [inline source/SpinalHDL/Libraries/stream.rst:494]:
```suggestion

A mux implementation for stream, which connects output to input specified by `io.select` signal.
```

```suggestion

A mux implementation for ``Stream``. It takes a ``select: UInt`` signal and an ``inputStreams: Vec(Stream)``, and returns a ``Stream`` which is connected to the input specified by ``select``.
```

File: source/SpinalHDL/Libraries/stream.rst [hunk 5]
```diff
@@ -489,6 +489,29 @@ or
...
+   val outputStream = StreamMux(select, inputStreams)
+
+.. note::
+   The `UInt` type of `select` signal could not be changed while output stream is stalled, or it might break the transaction on the fly.
+   Use `Stream` typed `select` can generate a stream interface which only fire and change the routing when it is safe.
+
+
+StreamDemux
+^^^^^^^^^^
+A demux for stream which route input to output stream specified by `select`. Refer the notes above.
```
Comment (numero-744) [inline source/SpinalHDL/Libraries/stream.rst:508]:
```suggestion
A demux for stream which routes input to output stream specified by `select`. Refer the notes above.
```

```suggestion
A demux implementation for ``Stream``. It takes an ``inputStream: Stream``, a ``select: UInt`` and a ``portCount: Int`` and returns a ``Vec(Stream)`` where the output stream specified by ``select`` is connected to ``inputStream``, the other output streams are inactive. Refer the notes above.
```

**Review by numero-744 — CHANGES_REQUESTED** [review]
Thanks for keeping RTD up-to-date with your contributions :smile:

File: source/SpinalHDL/Libraries/stream.rst (hunk 4 above)
Comment (Readon) [reply source/SpinalHDL/Libraries/stream.rst:494]: I think the existing style of documentation is more about to show the code by example but not an API Reference.
So the description of those arguments is really not that necessary to show the usage, especially on their types.

On other side, if there is something special need to be mentioned, then it is necessary to explain the example with signal in it.
There are examples explain that begin with "In above example, ....".

There should be a common policy there.

Comment (numero-744) [reply source/SpinalHDL/Libraries/stream.rst]: Yes, just fixing the typo might be enough (first suggestion). What I didn't understand after first read was the "connect output to input specified by `io.select` signal.": in the example I don't see a `Component` so I cannot find an `io.select`. This I why I suggested the version which is more "arguments to returned value".

> There should be a common policy there.

True

**Comment by numero-744** [comment]
LGTM (I cannot merge)

File: source/SpinalHDL/Libraries/stream.rst (hunk 4 above)
Comment (Readon) [reply source/SpinalHDL/Libraries/stream.rst]: > 

emmm, the io.select is in the StreamMux's definition.
So whether use a io signals or signals in example to explain is a problem.

**Comment by Dolu1990** [comment]
@numero-744 i just added rights for you to merge in this repo. Enjoy it ^^

**Review by numero-744 — APPROVED** [review]

**Pull request merged**